chore: add best practices for styling a component in codebase (#48808)
freeCodeCamp/freeCodeCamp: docs/codebase-best-practices.md

# Codebase Best Practices

## Styling a component

Be mindful of adding new colors to the codebase, as you will have to account for contrast and color blindness. Because of this, we always use already determined variables in [variable.css](/client/src/components/layouts/variables.css).

> We are against adding new variables to [variable.css](/client/src/components/layouts/variables.css), because we have to account for contrast, color blindness 🙂, and cluttering the devtools which affects negatively the devtools experiences and add concerns for accessibility.

Reviewer: These first two paragraphs seem like they're kind of saying the same thing. Maybe combine them into one? I would maybe just mention to use the colors from the variables.css file - and you could maybe link to the design style guide.

> We recommend using our [design style guide](https://design-style-guide.freecodecamp.org/) in styling components, you will find the colors mentioned in it in our [variable.css](/client/src/components/layouts/variables.css) file, and will find the fonts in [fonts.css](/client/src/components/layouts/fonts.css) file.
> 
> We are against adding new variables to [variable.css](/client/src/components/layouts/variables.css), because the new colors have to account for contrast, color blindness, and cluttering the devtools which affects negatively the developing experiences and add concerns for accessibility.
> 
> Use `!important` for accessibility concerns only after leaving a comment describing the issue, so it won't be removed by mistake in future refactoring.

Reviewer 2: ```suggestion
We recommend styling components using our [design style guide](https://design-style-guide.freecodecamp.org/). 

The colors are defined in [`variable.css`](/client/src/components/layouts/variables.css), and the fonts are in [`fonts.css`](/client/src/components/layouts/fonts.css).

We are strongly opinionated about adding new variables/tokens to the colors. After careful research, the colors have been chosen to respect the freeCodeCamp brand identity, developer experience, and accessibility.

The `!important` keyword may be used to override values in some cases (for ex.: accessibility concerns). You should add a comment describing the issue, so it doesn't get removed in future refactoring.
```

Break it up, make it clear and use a direct tone.

### RTL support

> We are striving to support RTL layout in the codebase, for this you need be mindful of how to style the component in your PR. Here are a quick role of thumbs to follow:

Reviewer: ```suggestion
We are striving to support right-to-left (RTL) layout in the codebase for languages that are read in this direction. For this you need be mindful of how to style components. Here are a quick role of thumbs to follow:
```

> - Don't use float properties: Although it may seem best to have the component in the "perfect position", you will be climbing mountains to reach that perfect positioning in responsive layout, and you will need to reach higher heights to have it supported in RTL. 

Reviewer: ```suggestion
- Don't use `float` properties: Although it may seem best to have the component in the "perfect position", you will be climbing mountains to reach that perfect positioning in responsive layout, and you will need to reach higher heights to have it supported in RTL. 
```

> - - Use flexbox and Grid layouts instead, as they have RTL support already built in them, and it will be easier to maintain and review

Reviewer: ```suggestion
- - Use flexbox and Grid layouts instead, as they have RTL support already built-in, and those will be easier to maintain and review
```

> - Don't define the direction while using `margin` and `padding`: it may seem harmless to use `padding-right` and `margin-left`, but these directions aren't mirrored when the layout change to RTL, and adding counter values for them in the RTL file makes maintaining the codebase harder.

Reviewer: ```suggestion
- Don't define the direction while using `margin` and `padding`: it may seem harmless to use `padding-right` and `margin-left`, but these directions aren't mirrored when the layout changes to RTL, and adding counter values for them in the RTL file makes maintaining the codebase harder.
```

We recommend styling components using our [design style guide](https://design-style-guide.freecodecamp.org/). 

The colors are defined in [`variable.css`](/client/src/components/layouts/variables.css), and the fonts are in [`fonts.css`](/client/src/components/layouts/fonts.css).

We are strongly opinionated about adding new variables/tokens to the colors. After careful research, the colors have been chosen to respect the freeCodeCamp brand identity, developer experience, and accessibility.

The `!important` keyword may be used to override values in some cases (for ex.: accessibility concerns). You should add a comment describing the issue, so it doesn't get removed in future refactoring.

### RTL support

We are striving to support right-to-left (RTL) layout in the codebase for languages that are read in this direction. For this you need be mindful of how to style components. Here are a quick role of thumbs to follow:

- Don't use `float` properties: Although it may seem best to have the component in the "perfect position", you will be climbing mountains to reach that perfect positioning in responsive layout, and you will need to reach higher heights to have it supported in RTL. 

> - - Use flexbox and Grid layouts instead, as they have RTL support already built-in, and those will be easier to maintain and review

Author: ```suggestion
- - Use Flexbox and Grid layouts instead, as they have RTL support already built-in, and those will be easier to maintain and review.
```
